### PR TITLE
fix(PluginDefinitions): Disallow defaulting of Option with Type Secret

### PR DESCRIPTION
**Pull request by uwe-mayer**
## Description

This PR fixes https://github.com/cloudoperators/greenhouse/issues/195

## What type of PR is this? (check all applicable)

- [ ] 🍕 Feature
- [x] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [x] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Added tests?

- [x] 👍 yes
- [ ] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help
- [ ] Separate ticket for tests # (issue/pr)

### CODE REVIEW

File: pkg/apis/greenhouse/v1alpha1/plugindefinition_types.go [hunk 1]
```diff
@@ -111,6 +111,15 @@ func (p *PluginOption) IsValid() error {
 	case PluginOptionTypeMap:
 		var m map[string]any
 		return json.Unmarshal(p.Default.Raw, &m)
+	case PluginOptionTypeSecret:
+		var s string
+		if err := json.Unmarshal(p.Default.Raw, &s); err != nil {
+			return err
+		}
+		if s != "" {
```
Comment (IvoGoman) [inline pkg/apis/greenhouse/v1alpha1/plugindefinition_types.go:119]: It is sufficient to check for the length of `Default.Raw` to make sure the value was not set
```suggestion
		if len(p.Default.Raw) != 0 {
```

File: pkg/admission/plugindefinition_webhook.go [hunk 2]
```diff
@@ -93,7 +93,7 @@ func validatePluginDefinitionOptionValueAndType(pluginDefinition *greenhousev1al
 		if err := option.IsValid(); err != nil {
 			return apierrors.NewInvalid(pluginDefinition.GroupVersionKind().GroupKind(), pluginDefinition.GetName(), field.ErrorList{
 				field.Invalid(field.NewPath("spec").Child("options").Child("name"), option.Name,
-					"A PluginOption Default must match the specified Type."),
+					"A PluginOption Default must match the specified Type and no defaults are allowed in PluginOptions of Type Secret."),
```
Comment (IvoGoman) [inline pkg/admission/plugindefinition_webhook.go:96]: NIT: Just a suggestion for the wording.
```suggestion
					"A PluginOption Default must match the specified Type, and defaults are not allowed in PluginOptions of the 'Secret' type."),
```

**Review by IvoGoman — CHANGES_REQUESTED** [review]
One remark to checking if the Raw JSON is empty for the Secret type.

File: pkg/apis/greenhouse/v1alpha1/plugindefinition_types.go (hunk 1 above)
Comment (uwe-mayer) [reply pkg/apis/greenhouse/v1alpha1/plugindefinition_types.go:119]: :heavy_check_mark:

**Review by IvoGoman — APPROVED** [review]

File: pkg/apis/greenhouse/v1alpha1/plugindefinition_types.go (hunk 1 above)
Comment (uwe-mayer) [reply pkg/apis/greenhouse/v1alpha1/plugindefinition_types.go]: Hmm tests fail after this merge.

The `Raw` representation for 
- `nil` is `[]uint8 len: 4, cap: 8, [110,117,108,108]` (a.k.a. `"null"`)   
- `""` is  `[]uint8 len: 2, cap: 8, [34,34]`  (a.k.a. `"\"\""`)

Both `len(p.Default.Raw)>0` if we do not unmarshal the json first. I think this was the reason why I added the unmarshalling in the first place...

I added a second (also failing) test to illustrate. 
We could stop using `json.Marshal(val)` for `val==nil` in the tests, which is done by the `test.MustReturnJSONFor()` func.
But I would prefer to catch the wider range of _not defaulted_ string values?

Comment (IvoGoman) [reply pkg/apis/greenhouse/v1alpha1/plugindefinition_types.go]: My Bad, I had just tested this with by converting `""` to `[]byte` and that looked good. Yeah let's do it the way you have proposed intiially, then we are on the safe side.

**Pull request merged**